cmd/snap/cmd_connections.go

```diff
@@ -0,0 +1,222 @@
+		return &cmdConnections{}
+	}, map[string]string{
+		"all":          i18n.G("Show connected and unconnected plugs and slots"),
+		"disconnected": i18n.G("Show disconnected plugs and slots"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: i18n.G("<snap>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Constrain listing to a specific snap"),
+	}})
+}
+
+func isSystemSnap(snap string) bool {
+	return snap == "core" || snap == "snapd" || snap == "system"
+}
+
+func endpoint(snap, name string) string {
+	if isSystemSnap(snap) {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+func wantedSnapMatches(name, wanted string) bool {
+	if wanted == "system" {
+		switch name {
+		case "core", "snapd", "system":
+			return true
+		default:
+			return false
+		}
+	}
+	return wanted == name
+}
+
+type connectionNotes struct {
+	slot      string
+	plug      string
+	manual    bool
+	gadget    bool
+	undesired bool
+}
+
+func (cn connectionNotes) String() string {
+	opts := []string{}
+	if cn.undesired {
+		opts = append(opts, "disconnected")
+	}
+	if cn.manual {
+		opts = append(opts, "manual")
+	}
+	if cn.gadget {
+		opts = append(opts, "gadget")
+	}
+	if len(opts) == 0 {
+		return "-"
+	}
+	return strings.Join(opts, ",")
+}
+
+func connName(conn client.Connection) string {
+	return endpoint(conn.Plug.Snap, conn.Plug.Name) + " " + endpoint(conn.Slot.Snap, conn.Slot.Name)
+}
+
+func (x *cmdConnections) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	opts := client.ConnectionOptions{
+		All: x.All || x.Disconnected,
+	}
+	wanted := string(x.Positionals.Snap)
+	if wanted != "" {
+		// when asking for a single snap, include its disconnected plugs
+		// and slots
+		opts.Snap = wanted
+		opts.All = true
+		// print all slots, unless we were asked for disconnected ones
+		// only
+		x.All = !x.Disconnected
+	}
+
+	connections, err := x.client.Connections(&opts)
+	if err != nil {
+		return err
+	}
+	if len(connections.Plugs) == 0 && len(connections.Slots) == 0 {
+		return fmt.Errorf(i18n.G("no connections found"))
+	}
+
+	notes := make(map[string]connectionNotes, len(connections.Established)+len(connections.Undesired))
+	for _, conn := range connections.Established {
+		notes[connName(conn)] = connectionNotes{
+			plug:      endpoint(conn.Plug.Snap, conn.Plug.Name),
+			slot:      endpoint(conn.Slot.Snap, conn.Slot.Name),
+			manual:    conn.Manual,
+			gadget:    conn.Gadget,
+			undesired: false,
+		}
+	}
+	for _, conn := range connections.Undesired {
+		notes[connName(conn)] = connectionNotes{
+			plug:      endpoint(conn.Plug.Snap, conn.Plug.Name),
+			slot:      endpoint(conn.Slot.Snap, conn.Slot.Name),
+			manual:    conn.Manual,
+			gadget:    conn.Gadget,
+			undesired: true,
+		}
+	}
+
+	w := tabWriter()
+	fmt.Fprintln(w, i18n.G("Plug\tSlot\tInterface\tNotes"))
```

Comment:
If a snap only has connected and the user asked for disconnected, we'll print the header and then nothing more.

Reply:
Fixed